Plutarch/DataRepr/Internal.hs

```diff
@@ -180,7 +184,55 @@ type family PUnLabel (a :: PLabeledType) :: PType where
  This is how most data structures are stored on-chain.
 -}
 type PDataSum :: [[PLabeledType]] -> PType
-data PDataSum (defs :: [[PLabeledType]]) (s :: S)
+data PDataSum (defs :: [[PLabeledType]]) (s :: S) where 
+  DSZ :: (forall {out :: PType}. (Term s (PDataRecord x) -> Term s out) -> Term s out) 
+      -> PDataSum (x ': xs) s
+
+  DSS :: PDataSum xs s -> PDataSum (x ': xs) s
```

Comment:
This is now isomorphic to `NS`, right? viz.:

```haskell
newtype DRec s x = DRec (forall {out :: PType}. (Term s (PDataRecord x) -> Term s out) )
type PDataSum defs s = NS (DRec s) defs
```

Reply:
Yeah, I realized that like 20 seconds after the last commit. Should be an easy change.

Reply:
`PDataSum` has to be a newtype to get the instance to work (GHC restriction on partially applied type synonyms) and I think `DRec` has to be a little different: 

```haskell
newtype DRec s x = DRec (forall {out :: PType}. ((Term s (PDataRecord x) -> Term s out) -> Term s out))

type PDataSum :: [[PLabeledType]] -> PType
newtype PDataSum defs s = PDataSum (NS (DRec s) defs)
```
Morally I think those are close enough to what you suggested though.